### PR TITLE
Decouple direct access to native::scalar_tensor from TensorIndexing.h

### PR DESCRIPTION
**Pull request by iseeyuan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48761 Decouple direct access to native::scalar_tensor from TensorIndexing.h**

Targeting one of the items in https://github.com/pytorch/pytorch/issues/48684. For performance purpose we don't use at::scalar_tensor. Since scalar_tensor_static is available for CPU we could use it at least for CPU. One uncertainty is the CUDA performance. But there's no fast path for CUDA under native::scalar_tensor either, I assume the perf on CUDA may not be affected.

Differential Revision: [D25410975](https://our.internmc.facebook.com/intern/diff/D25410975)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Optimistically approving.

**Comment by ezyang** [comment]
hmm, looks like tests are failing though

**Comment by iseeyuan** [comment]
> hmm, looks like tests are failing though

Looking

File: aten/src/ATen/TensorIndexing.h [hunk 1]
```diff
@@ -320,8 +317,11 @@ static inline int64_t count_specified_dimensions(const ArrayRef<TensorIndex>& in
 // The rest of the functions are in `at::indexing::impl` namespace, signifying
 // that they shouldn't be used from Python indexing implementation.
 static inline Tensor scalarToTensor(Scalar v, const TensorOptions& options, const at::Device& self_device) {
-  if (self_device == at::kCPU || self_device == at::kCUDA) {
-    return impl::scalarToTensorCPUOrCUDA(v, options);
+  if (self_device == at::kCPU && !v.isComplex() &&
+      options.dtype_opt()->toScalarType() != ScalarType::ComplexDouble &&
```
Comment (iseeyuan) [inline aten/src/ATen/TensorIndexing.h:321]: @ezyang , the CI failures were caused majorly by complex types in the options. After filtering them out CI looks fine now.

Comment (ezyang) [reply aten/src/ATen/TensorIndexing.h:321]: The condition here looks kind of beefy now. Do you know why you need to filter out complex here? That seems pretty unexpected...

Comment (iseeyuan) [reply aten/src/ATen/TensorIndexing.h:321]: There would be some unit test failure if we don't filter them out. For example: 
```
======================================================================
ERROR: test_autograd_simple_views_python (__main__.TestAutograd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_autograd.py", line 4519, in test_autograd_simple_views_python
    self._do_test_autograd_simple_views_python(torch.cdouble)
  File "test/test_autograd.py", line 4488, in _do_test_autograd_simple_views_python
    gradcheck(fn, (a, b))
  File "/Users/myuan/src/pt_1/pytorch/torch/autograd/gradcheck.py", line 343, in gradcheck
    nondet_tol=nondet_tol)
  File "/Users/myuan/src/pt_1/pytorch/torch/autograd/gradcheck.py", line 181, in get_analytical_jacobian
    flat_grad_output[i] = grad_out
RuntimeError: "fill_out" not implemented for 'ComplexDouble'
```

Comment (ezyang) [reply aten/src/ATen/TensorIndexing.h:321]: But this just sounds like we should actually implement fill_out on complex, and then you don't need this condition anymore...

**Pull request closed without merging**